contributors/design-proposals/storage/cloning.md

```diff
@@ -0,0 +1,158 @@
+* An admin wishes to create a copy of a volume to spawn another container to use the same data. 
+* An admin wishes to create a copy of a volume to pre-seed a process with information the container might need for bootstrapping.
+* A user wishes to clone a volume for purposes of disaster recovery
+
+## Design Overview
+The proposed design is based on the idea of leveraging well defined concepts for storage in Kubernetes. The cloning will be initiated 
+by creating an PVC to create a new volume using the dataSource API created with snapshots.
+The actual cloning process must be implemented by the provisioning storage and properly understand and react to the request to clone. 
+
+### API
+Leveraging the proposed API change for 'dataSource' the PVC will make a clone request by specifying the object it wishes to clone. In this example, we are cloning a pvc. For version 1 of cloning, we will limit cloning of pvcs to the current namespace only.
+
+** Pre-Request:
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: pv-1
+  Namespace: myns
+spec:
+  capacity:
+    storage: 100Gi
+  volumeMode: Filesystem
+accessModes:
+    - ReadOnlyMany
+  persistentVolumeReclaimPolicy: Delete
+  storageClassName: csi-gce-pd
+```
+```
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: pvc-1
+  Namespace: myns
+spec:
+  capacity:
+    storage: 100Gi
+```
+
+** Request:
+```
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: clone-pvc
+  Namespace: myns
+spec:
+  accessModes:
+    - ReadWriteOnce
+  storageClassName: csi-gce-pd
+  dataSource:
+    kind: PersistentVolume
+    name: pv-1
+  resources:
+    requests:
+      storage: 10Gi
+```
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: csi-gce-pd
+provisioner: kubernetes.io/gce-pd
+```
+
+** Result:
+```
+kind: PersistentVolume
+apiVersion: v1
+metadata:
+  name: pv-2
+  Namespace: myns
+spec:
+  accessModes:
+    - ReadWriteOnce
+  storageClassName: csi-gce-pd
+  resources:
+    requests:
+      storage: 10Gi
+```
+```
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: clone-pvc
+  Namespace: myns
+spec:
+  accessModes:
+    - ReadWriteOnce
+  storageClassName: csi-gce-pd
+  dataSource:
+    kind: PersistentVolume
+    name: pv-1
+  resources:
+    requests:
+      storage: 10Gi
+```   
+
+### CSI Spec
+This PR does not propose an API change, however the CSI specification will need to be updated as follows to add an additional
+option for the VolumeContentSource:
+
+``` 
+// Specifies what source the volume will be created from. One of the
+// type fields MUST be specified.
+message VolumeContentSource {
+  message SnapshotSource {
+    // Contains identity information for the existing source snapshot.
+    // This field is REQUIRED. Plugin is REQUIRED to support creating
+    // volume from snapshot if it supports the capability
+    // CREATE_DELETE_SNAPSHOT.
+    string id = 1;
+  }
+  message PersistentVolumeSource {
```

Comment:
This ones under review, uses a more generic term of "VolumeSource" which I think is preferable.
https://github.com/container-storage-interface/spec/pull/244